step-api/src/main/java/org/jenkinsci/plugins/workflow/steps/AbstractSynchronousNonBlockingStepExecution.java

```diff
@@ -0,0 +1,84 @@
+package org.jenkinsci.plugins.workflow.steps;
+
+import hudson.util.DaemonThreadFactory;
+import hudson.util.NamingThreadFactory;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+/**
+ * Similar to {@link AbstractSynchronousStepExecution} but it always executes synchronously and does not block the CPS VM thread.
+ * @see {@link StepExecution}
+ * @param <T> the type of the return value (may be {@link Void})
+ */
+public abstract class AbstractSynchronousNonBlockingStepExecution<T> extends AbstractStepExecutionImpl {
+
+    private transient volatile Future<?> task;
+
+    protected AbstractSynchronousNonBlockingStepExecution() {
+    }
+
+    protected AbstractSynchronousNonBlockingStepExecution(StepContext context) {
+        super(context);
+    }
+
+    /**
+     * Meat of the execution.
+     *
+     * When this method returns, a step execution is over.
+     */
+    protected abstract T run() throws Exception;
+
+    @Override
+    public final boolean start() throws Exception {
+        task = StepRunnerExecutorService.get().submit(new StepRunner(this));
+        return false;
+    }
+
+    /**
+     * If the computation is going synchronously, try to cancel that.
+     */
+    @Override
+    public void stop(Throwable cause) throws Exception {
+        if (task != null) {
+            task.cancel(true);
+        }
+        getContext().onFailure(cause);
+    }
+
+    @Override
+    public void onResume() {
+        getContext().onFailure(new Exception("Resume after a restart not supported for non-blocking synchronous steps"));
+    }
+
+    private static class StepRunner implements Runnable {
+
+        private final AbstractSynchronousNonBlockingStepExecution<?> step;
+
+        public StepRunner(AbstractSynchronousNonBlockingStepExecution<?> step) {
+            this.step = step;
+        }
```

Comment:
Well you can just write `getContext()` and the receiver is implicit. For `run()` it is true you need to disambiguate with `AbstractSynchronousNonBlockingStepExecution.this.`, part of the Java language since 1.1 but infrequently used. Guess it depends on what you are comfortable with.